test/Validate.SSA.cpp

```diff
@@ -0,0 +1,699 @@
+%vec3   = OpTypeVector %intt 3
+%vec4   = OpTypeVector %intt 4
+%mat33  = OpTypeMatrix %vec3 3
+%mat44  = OpTypeMatrix %vec4 4
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardDecorateInvalidIdBad) {
+  char str[] = R"(
+          OpMemoryModel Logical GLSL450
+          OpEntryPoint GLCompute %3 ""
+          OpExecutionMode %3 LocalSize 1 1 1
+          OpDecorate %missing Restrict
+%voidt  = OpTypeVoid
+%intt   = OpTypeInt 32 1
+%ptrt   = OpTypePointer UniformConstant %intt
+%var    = OpVariable %ptrt UniformConstant
+%2      = OpTypeFunction %voidt
+%3      = OpFunction %voidt None %2
+%4      = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardFunctionCallGood) {
+  char str[] = R"(
+        OpMemoryModel Logical GLSL450
+        OpEntryPoint GLCompute %5 ""
+        OpExecutionMode %5 LocalSize 1 1 1
+%1    = OpTypeVoid
+%2    = OpTypeInt 32 1
+%3    = OpTypeInt 32 0
+%4    = OpTypeFunction %1
+%5    = OpFunction %1 None %4
+%6    = OpLabel
+%four = OpConstant %2 4
+%five = OpConstant %3 5
+%7    = OpFunctionCall %1 %9 %four %five
+        OpFunctionEnd
+%8    = OpTypeFunction %1 %2 %3
+%9    = OpFunction %1 None %8
+%10   = OpFunctionParameter %2
+%11   = OpFunctionParameter %3
+%12   = OpLabel
+        OpReturn
+        OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, DISABLED_ForwardBranchConditionalGood) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+%true   = OpConstantTrue %boolt
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %truel %falsel
+%truel  = OpLabel
+          OpNop
+          OpBranch %endl
+%falsel = OpLabel
+          OpNop
+%endl    = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, DISABLED_ForwardBranchConditionalWithWeightsGood) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+%true   = OpConstantTrue %boolt
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %truel %falsel !1 !9
+%truel  = OpLabel
+          OpNop
+          OpBranch %endl
+%falsel = OpLabel
+          OpNop
+%endl   = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, ForwardBranchConditionalNonDominantConditionBad) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %missing %falsel
+%truel  = OpLabel
+          OpNop
+          OpBranch %endl
+%falsel = OpLabel
+          OpNop
+%endl   = OpLabel
+%true   = OpConstantTrue %boolt
```

Comment:
This will fail in the future because the constants should all be defined before any function bodies.
Better to change this line to 
```
%true_copy = OpCopyObject %true
```
and change the previous OpBranchConditional to use %true_copy